lib/cc/cli/test.rb

```diff
@@ -0,0 +1,233 @@
+
+module CC
+  module CLI
+    class Test < Command
+      Marker = Struct.new(:line, :line_text, :issue)
+
+      def run
+        @engine_name = @args.first
+
+        if @engine_name.blank?
+          fatal "Usage: codeclimate test #{rainbow.wrap("engine_name").underline}"
+        end
+
+        tmpdir = create_tmpdir
+
+        Dir.chdir(tmpdir) do
+          prepare_working_dir
+
+          test_paths.each do |test_path|
+            unpack(test_path)
+          end
+
+          Dir["**/*"].each do |test_file|
+            next unless File.file?(test_file)
+            process_file(test_file, tmpdir)
+          end
+        end
+      ensure
+        remove_null_container
+      end
+
+      def process_file(test_file, tmpdir)
+        markers = markers_in(test_file)
+
+        actual_issues = issues_in(test_file)
+        compare_issues(actual_issues, markers)
+      end
+
+      def compare_issues(actual_issues, markers)
+        markers.each do |marker|
+          validate_issue(marker, actual_issues)
+        end
+
+        validate_unexpected_issues(actual_issues)
+      end
+
+      def validate_issue(marker, actual_issues)
+        found_index = nil
+
+        actual_issues.each_with_index do |actual, index|
+          if fuzzy_match(marker.issue, actual)
+            found_index = index
+            break
+          end
+        end
+
+        if found_index
+          say "PASS %3d: %s" % [marker.line, marker.line_text]
+          actual_issues.delete_at(found_index)
+        else
+          say colorize("MISS %3d: %s" % [marker.line, marker.line_text], :red)
+          say colorize("Searched:", :yellow)
+          say colorize(JSON.pretty_generate(marker.issue), :yellow)
+          say "\n"
+          say colorize("Actual:", :yellow)
+          say colorize(JSON.pretty_generate(actual_issues), :yellow)
+          fatal "Expected issue not found."
+        end
+      end
+
+      def validate_unexpected_issues(actual_issues)
+        if actual_issues.any?
+          say colorize("Actuals not empty after matching.", :red)
+          say
+          say colorize("#{actual_issues.size } remaining:", :yellow)
+          say colorize(JSON.pretty_generate(actual_issues), :yellow)
+          fatal "Unexpected issue found."
+        end
+      end
+
+      def fuzzy_match(expected, actual)
+        expected.all? do |key, value|
+          actual[key] == value
+        end
+      end
+
+      def issues_in(test_file)
+        issue_docs = capture_stdout do
+          codeclimate_analyze(test_file)
+        end
+
+        JSON.parse(issue_docs)
+      end
+
+      def codeclimate_analyze(relative_path)
+        codeclimate_path = File.expand_path(File.join(File.dirname(__FILE__), "../../../bin/codeclimate"))
+
+        system([
+          "unset CODE_PATH &&",
+          "unset FILESYSTEM_DIR &&",
+          codeclimate_path,
+          "analyze",
+          "--engine", @engine_name,
+          "-f", "json",
+          relative_path
```

Comment:
I think you probably want this quoted in case of spaces. (Possibly ditto for the `codeclimate_path`)